src/cache_padded.rs

```diff
@@ -1,120 +1,84 @@
-use std::marker;
-use std::cell::UnsafeCell;
 use std::fmt;
+use std::marker::PhantomData;
 use std::mem;
-use std::ptr;
 use std::ops::{Deref, DerefMut};
+use std::ptr;
 
-// For now, treat this as an arch-independent constant.
-const CACHE_LINE: usize = 32;
+/// 64 bytes is the most common cache line size on modern machines.
+const CACHE_LINE_BYTES: usize = 64;
 
-#[cfg_attr(feature = "nightly", repr(simd))]
-#[derive(Debug)]
-struct Padding(u64, u64, u64, u64);
+#[cfg(not(feature = "nightly"))]
+struct Inner<T> {
+    bytes: [u8; CACHE_LINE_BYTES],
+    _marker: PhantomData<T>,
+}
 
-/// Pad `T` to the length of a cacheline.
-///
-/// Sometimes concurrent programming requires a piece of data to be padded out
-/// to the size of a cacheline to avoid "false sharing": cachelines being
-/// invalidated due to unrelated concurrent activity. Use the `CachePadded` type
-/// when you want to *avoid* cache locality.
+#[cfg(feature = "nightly")]
+#[repr(align(64))]
+#[allow(unions_with_drop_fields)]
+union Inner<T> {
+    bytes: [u8; CACHE_LINE_BYTES],
+    _value: T,
+}
+
+/// Pads `T` to the length of a cache line.
 ///
-/// At the moment, cache lines are assumed to be 32 * sizeof(usize) on all
-/// architectures.
+/// Sometimes concurrent programming requires a piece of data to be padded out to the size of a
+/// cacheline to avoid "false sharing": cache lines being invalidated due to unrelated concurrent
+/// activity. Use this type when you want to *avoid* cache locality.
 ///
-/// **Warning**: the wrapped data is never dropped; move out using `ptr::read`
-/// if you need to run dtors.
+/// At the moment, cache lines are assumed to be 64 bytes on all architectures.
+/// Note that, while the size of `CachePadded<T>` is 64 bytes, the alignment may not be.
+#[allow(unions_with_drop_fields)]
 pub struct CachePadded<T> {
-    data: UnsafeCell<[usize; CACHE_LINE]>,
-    _marker: ([Padding; 0], marker::PhantomData<T>),
-}
+    inner: Inner<T>,
 
-impl<T> fmt::Debug for CachePadded<T> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "CachePadded {{ ... }}")
-    }
+    /// `[T; 0]` ensures correct alignment.
```

Comment:
Is it a documented behavior?

Reply:
I didn't find anything explicit on this issue, but FWIW, there's this: [link](https://github.com/rust-lang/rfcs/issues/325#issuecomment-305046766)